### PR TITLE
[SPARK-11313][SQL] implement cogroup on DataSets

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 1]
```diff
@@ -513,3 +513,16 @@ case class MapGroups[K, T, U](
   override def missingInput: AttributeSet = AttributeSet.empty
 }
 
+case class CoGroup2[K, T, U, V](
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Consider adding a factory like in the other cases to simplify implicit passing.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Also I'd probably just call it cogroup as we'll probably use a single variadic operator if we decide to do more than 2.

**Comment by marmbrus** [comment]
Looking pretty good so far!  Let's wait to add cogroup for more than 2 datasets.

**Comment by cloud-fan** [comment]
cc @marmbrus

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
will open it again when we need to support cogroup on more than 2 datasets.

**Pull request closed without merging**